File: javascript/life.js
#! /usr/bin/env node

// pretend we can be functional: use immutable data types
const { Set, Map } = require('immutable')
const live = "live";
const die = "die";
const showWork = false;
const generations = 1000;

// build an immutable cell position value
function pos(x, y) {
    return new Map({"x":x, "y":y})
}

// build an immutable cell change value
function change(liveDie, pos) {
    return new Map({"change": liveDie, "pos":pos})
}

// build a board, with live cells and next set of updates
function board(liveSet, updates) {
    return new Map({"liveSet":liveSet, "updates": updates});
}

// doesn't return an immutable value, but is only used for display
function boundingBox(liveSet) {
    if (liveSet.isEmpty()) {
	return [[0,0], [1,1]];
    }
    const xs = liveSet.map(p => p.get("x")).toArray();
    const ys = liveSet.map(p => p.get("y")).toArray();
    return [[Math.min(...xs), Math.min(...ys)],
	    [Math.max(...xs), Math.max(...ys)]]
}

// a simple range function
function range(start, end) {
    return [...Array(end - start).keys()].map(i => i + start);
}

// gross writing to stdout with console.log, but I don't know better
function clearScreen() {
    console.log("\033[2J\033[;H")
}

function printBoard(liveSet) {
    clearScreen();
    const bbox = boundingBox(liveSet);
    range(bbox[0][1], bbox[1][1] + 1).reverse().map(y => {
	const row = range(bbox[0][0], bbox[1][0] + 1);
	console.log(row.map(x => liveSet.has(pos(x, y)) ? "#" : " ").join(""));
    });
}

function applyUpdates(current, updates) {
    const toLive = new Set(
	updates.filter(e => e.get("change") == live).map(e => e.get("pos"))
    );
    const toDie = new Set(
	updates.filter(e => e.get("change") == die).map(e => e.get("pos"))
    );
    return current.union(toLive).subtract(toDie);
}

// This isn't functional: replacements wanted
function eight(p) {
    const result = []
    const px = p.get("x")
    const py = p.get("y")
    for (x = -1; x <= 1; x++) {
	for (y = -1; y <= 1; y++) {
	    if (x != 0 || y != 0) {
		result.push(pos(px + x, py + y));
	    }
	}
    }
    return result;
}

function neighbors(updates) {
    const each = updates.map(c => c.get("pos")).map(p => eight(p))
    return new Set([].concat(...each));
}

function neighborCount(alive, pos) {
    return eight(pos).map(p => alive.has(p)).filter(v => v).length;
}

function computeChange(liveSet, pos) {
    const count = neighborCount(liveSet, pos);
    if (count == 2) {
	return null;
    }
    const isAlive = liveSet.has(pos);
    if (count == 3) {
	if (isAlive) {
	    return null;
	} else {
	    return change("live", pos);
	}
    }
    if (isAlive) {
	return change("die", pos);
    }
    return null;
}

function computeChanges(alive, affected) {
    return affected.map(p => computeChange(alive, p)).filter(c => c);
}

function generation(count, b) {
    if (count <= 0) {
	return;
    }
    const liveSet = b.get("liveSet");
    const updates = b.get("updates");
    const newLiveSet = applyUpdates(liveSet, updates);
    const affected = neighbors(updates)
    const newUpdates = computeChanges(newLiveSet, affected)
    if (showWork) {
	printBoard(newLiveSet);
	setTimeout(() => generation(count - 1, board(newLiveSet, newUpdates)),
		   1000 / 30);
    } else {
	generation(count - 1, board(newLiveSet, newUpdates));
    }
}

const r_pentomino = [[0, 0], [0, 1], [1, 1], [-1, 0], [0, -1]]

function main() {
    const updates = r_pentomino.map(a => change("live", pos(a[0], a[1])));
    const b = board(new Set(), updates);
    const start = Date.now();
    generation(generations, board(new Set(), updates));
    const millis = Date.now() - start;
    console.log(generations * 1000 / millis + " generations per second");
}

if (showWork) {
    main()
} else {
    main()
    main()
    main()
    main()
    main()
}

File: haskell/life.hs
import Data.Char (chr)
import Data.List (concat, intersperse)
import Control.Concurrent (threadDelay)
import qualified Data.Set as Set

humanSpeed = 1 / 30.0
generations = 1000
data State = Alive | Dead
data DestinyType = Live | Die | Ignore deriving (Show, Eq, Ord)
data Coord = Coord {coord_x :: Int, coord_y :: Int } deriving (Show, Eq, Ord)
type LiveSet = Set.Set(Coord)
type Change = (DestinyType, Coord)
type Board = (LiveSet, [Change])

clearScreen :: IO ()
clearScreen = do
  -- clear the screen
  putStr (esc : "[2J")
  -- move the cursor to the "home" position
  putStr (esc : "[;H")
  where
    esc = chr(27)

cell :: LiveSet -> Coord -> Char
cell liveSet coord =
  if (Set.member coord liveSet) then '@' else ' '

row :: LiveSet -> Int -> Int -> Int -> [Char]
row liveSet minx maxx y =
  map (\x -> cell liveSet (Coord x y)) [minx .. maxx]


expandBox :: ((Int, Int), (Int, Int)) -> Coord -> ((Int, Int), (Int, Int))
expandBox box coord =
  ((min minx x, min miny y), (max maxx x, max maxy y))
  where
    ((minx, miny), (maxx, maxy)) = box
    (x, y) = (coord_x coord, coord_y coord)

bbox :: Foldable t => t (Coord) -> ((Int, Int), (Int, Int))
bbox lst =
  foldl expandBox tinybox lst
  where
    minint = minBound::Int
    maxint = maxBound::Int
    tinybox = ((maxint, maxint), (minint, minint))

asString :: Board -> [Char]
asString board =
  concat (intersperse "\n"
          (map (row liveSet minx maxx) [maxy, maxy - 1 .. miny]))
  where
    (liveSet, updates) = board
    ((minx, miny), (maxx, maxy)) = bbox liveSet

printBoard :: Board -> IO ()
printBoard board = do
  clearScreen
  putStrLn (asString board)
  threadDelay (round (1000000 * humanSpeed))

applyUpdates :: LiveSet -> [Change] -> LiveSet
applyUpdates liveSet updates =
  Set.union (Set.difference liveSet toDie) toLive
  where
    toLive = Set.fromList [pos | (Live, pos) <- updates]
    toDie = Set.fromList [pos | (Die, pos) <- updates]

neighbors :: Coord -> [Coord]
neighbors pos =
  [Coord (posx + x) (posy + y)| x <- [-1 .. 1], y <- [-1 .. 1], x /= 0 || y /= 0]
  where
    (posx, posy) = (coord_x pos, coord_y pos)

considerSet :: [Coord] -> Set.Set(Coord)
considerSet positions =
  Set.fromList (concat (map neighbors positions))

-- rules of life
change :: Int -> State -> DestinyType
change 2 _     = Ignore
change 3 Dead  = Live
change 3 Alive = Ignore
change _ Alive = Die
change _ _     = Ignore

computeChange :: LiveSet -> Coord -> Change
computeChange liveSet pos =
  (how, pos)
  where
    neighborCount = length [True|n <- (neighbors pos), Set.member n liveSet]
    state = if Set.member pos liveSet then Alive else Dead
    how = change neighborCount state

computeUpdates :: LiveSet -> [Change] -> [Change]
computeUpdates liveSet updates =
  [(how, pos)|(how, pos) <- (Set.toList newUpdates), how /= Ignore]
  where
    changed = [pos|(how, pos) <- updates]
    newUpdates = Set.map (computeChange liveSet) (considerSet changed)

next :: Board -> Board
next board =
  (newSet, newUpdates)
  where
    (liveSet, updates) = board
    newSet = applyUpdates liveSet updates
    newUpdates = computeUpdates newSet updates

printGeneration :: Int -> Board -> IO ()
printGeneration 0 board = do
  -- do something with the last board to force evaluation
  -- otherwise nothing gets computed and we can't benchmark
  putStrLn (show (length updates))
  where
    (lifeSet, updates) = board

printGeneration n board =
  do
    -- for standard game-of-life display, uncomment this line:
    -- printBoard board
    printGeneration (n - 1) (next board)

run :: Int -> IO ()
run n = do
  printGeneration n start
  where
    r_pentomino = [(0, 0), (0, 1), (1, 1), (-1, 0), (0, -1)]
    update = map (\ (x, y) -> (Live, Coord x y)) r_pentomino
    liveSet = (Set.fromList [])
    start = (liveSet, update)

main :: IO ()
main = do
  run generations
  run generations
  run generations
  run generations
  run generations
